Introduce configuration options in the cluster API (#137)
valkey-io/libvalkey: docs/cluster.md

### Connecting

There are a few alternative ways to setup and connect to a cluster.
The basic alternatives lacks most options, but can be enough for some use cases.

```c
valkeyClusterContext *valkeyClusterConnect(const char *addrs);
valkeyClusterContext *valkeyClusterConnectWithTimeout(const char *addrs,
                                                      const struct timeval tv);

// There is also a convenience struct to specify various options.
valkeyClusterContext *valkeyClusterConnectWithOptions(const valkeyClusterOptions *options);
```

When connecting to a cluster, libvalkey will return `NULL` in the event that we can't allocate the context, and set the `err` member if we can allocate the context but there are issues.
So when connecting it's simple to handle error states.

```c
valkeyClusterContext *cc = valkeyClusterConnect("127.0.0.1:6379,127.0.0.1:6380");
if (cc == NULL || cc->err) {
    fprintf(stderr, "Error: %s\n", cc ? cc->errstr : "OOM");
}
```

### Connection options

> There are a variety of options you can specify when connecting to a cluster, which are delivered via the `valkeyClusterOptions` helper struct.

Reviewer: Delivered? :thinking: specified maybe? or provided, configured, ...

Helper struct? Just struct.

> This includes information how to connect to the cluster and defining optional callbacks as well as other flags.

Reviewer: Missing word "about".

"callbacks and other flags" sounds like callbacks are flags. But it's not only flags. It's password and other things. Let's just say "and other options"?

```suggestion
This includes information about how to connect to the cluster and defining optional callbacks and other options.
```

See [include/valkey/cluster.h](../include/valkey/cluster.h) for more details.

```c
valkeyClusterOptions opt = {0};
opt.initial_nodes = "127.0.0.1:6379,127.0.0.1:6380"; // Addresses to initially connect to.
opt.options = VALKEY_OPT_USE_CLUSTER_SLOTS;          // See available flags below.
opt.password = "password"                            // Authentication; libvalkey sends the `AUTH` command.

// There are helper functions to set some options.

> valkeyClusterOptionsEnableTLS(&opt, tlsCtx);         // Use TLS and a prepared `valkeyTLSContext` when connecting.

Reviewer: Why do we need this function to set TLS options? Can't you just set them on the opts object?

```c
opts.tls = tlsCtx;
```

Reviewer: Great. I hope you update the migration guide in parallel.

valkeyClusterContext *cc = valkeyClusterConnectWithOptions(&opt);
if (cc == NULL || cc->err) {
    fprintf(stderr, "Error: %s\n", cc ? cc->errstr : "OOM");
}
```

> There are also several flags you can specify when using the `valkeyClusterOptions` helper struct.

Reviewer: Is it `.flags`? Is it a bitwise-or?

```suggestion
There are also several flags you can specify in `valkeyClusterOptions.flags`. It's a bitwise or of the following flags:
```

valkeyClusterContext *valkeyClusterConnectWithTimeout(const char *addrs,
                                                      const struct timeval tv);
```

There is also a convenience struct to specify various options.

```c
valkeyClusterContext *valkeyClusterConnectWithOptions(const valkeyClusterOptions *options);
```

When connecting to a cluster, `NULL` is returned when the context can't be allocated, or `err` and `errstr` are set in the returned allocated context when there are issues.
So when connecting it's simple to handle error states.

```c
valkeyClusterContext *cc = valkeyClusterConnect("127.0.0.1:6379,127.0.0.1:6380");
if (cc == NULL || cc->err) {
    fprintf(stderr, "Error: %s\n", cc ? cc->errstr : "OOM");
}
```

### Connection options

There are a variety of options you can specify using the `valkeyClusterOptions` struct when connecting to a cluster.
This includes information about how to connect to the cluster and defining optional callbacks and other options.
See [include/valkey/cluster.h](../include/valkey/cluster.h) for more details.

```c
valkeyClusterOptions opt = {0};
opt.initial_nodes = "127.0.0.1:6379,127.0.0.1:6380"; // Addresses to initially connect to.

> opt.options = VALKEY_OPT_USE_CLUSTER_NODES;          // See available flags below.

Author: I used the name `options` for this field to match how it's named in `valkeyOptions` (which also has an `options` which takes `VALKEY_OPT_*`.).  I guess we should change both in that case.

When contexts are initiated we keep the given option-flags in a context-field named flags, but here we also use it to keep internal state, both in cluster contexts and standalone.

> opt.options = VALKEY_OPT_USE_CLUSTER_NODES;          // See available flags below.

Reviewer: We can keep options then. If we change to flags, it would imply all the macro names that contain VALKEY_OPT too. It would be too big change maybe.

> opt.options = VALKEY_OPT_USE_CLUSTER_NODES;          // See available flags below.

Author: Yes, it getting large then. We can think about it, and possibly replace them all in a separate PR afterwards.

opt.password = "password"                            // Authentication; libvalkey sends the `AUTH` command.

// Enable TLS using a prepared `valkeyTLSContext` when connecting.
options.tls = tlsCtx;
options.tls_init_fn = &valkeyInitiateTLSWithContext;

valkeyClusterContext *cc = valkeyClusterConnectWithOptions(&opt);
if (cc == NULL || cc->err) {
    fprintf(stderr, "Error: %s\n", cc ? cc->errstr : "OOM");
}
```

There are also several flags you can specify in `valkeyClusterOptions.flags`. It's a bitwise OR of the following flags:

| Flag | Description  |
| --- | --- |

> | VALKEY\_OPT\_USE\_CLUSTER\_NODES | Tells libvalkey to use the command `CLUSTER NODES` when updating its slot map (cluster topology).<br>Libvalkey uses `CLUSTER SLOTS` by default. |

Author: ah nice! That makes it readable when "raw-reading" from a terminal, will update standalone.md too.

Any command sent by the user thereafter will create a new non-blocking connection, unless a non-blocking connection already exists to the destination.
The function returns a pointer to a newly created `valkeyClusterAsyncContext` struct and its `err` field should be checked to make sure the initial slot map update was successful.

There is also a separate API to perform the context initiation and initial connect in separate steps.
This is useful when there is a need to provide an event callback with the current `valkeyClusterAsyncContext`.
The context is first initiated using `valkeyClusterAsyncContextInit` and then `valkeyClusterAsyncConnect` will initiate connection attempts.

```c
valkeyClusterOptions options = {
   .initial_nodes = "127.0.0.1:7000";
};
valkeyClusterOptionsUseLibev(&options, EV_DEFAULT);

// Initiate the context.
valkeyClusterAsyncContext *acc = valkeyClusterAsyncContextInit(&options);

// Set the event callback using the context as privdata.
valkeyClusterAsyncSetEventCallback(acc, eventCallback, acc);

// Start connecting to the initial nodes.
valkeyClusterAsyncConnect(acc)
```

### Connection options

> There are a variety of options you can specify using the `valkeyClusterOptions` struct when connecting to a cluster.

Reviewer: ```suggestion
There is a variety of options you can specify using the `valkeyClusterOptions` struct when connecting to a cluster.
```

Author: Corrected

> Any command sent by the user thereafter will create a new non-blocking connection, unless a non-blocking connection already exists to the destination.

Reviewer: (I can't comment on the lines above.) The async API doc starts like this:

> ## Asynchronous API
> ### Connecting
> There are two alternative ways to initiate a cluster client, which also determines how the client behaves during the initial connect.
The first alternative is to use the function `valkeyClusterAsyncConnect`, which initially connects to the cluster in a blocking fashion and waits for the slot map before returning.

... and then the variant with init + connect2 is described.

But I checked `cluster.h` and I think we deleted the blocking connect and renamed connect2 to connect. We don't have any blocking valkeyClusterAsyncConnect anymore, do we?

I think the doc should first describe this variant:

1. Create options struct
2. Attach event loop to options
3. valkeyClusterAsyncConnectWithOptions

```C
valkeyClusterOptions options = {
    .initial_nodes = "127.0.0.1:7000";
};
valkeyClusterOptionsUseLibev(&options, EV_DEFAULT);
valkeyClusterAsyncContext *acc = valkeyClusterAsyncConnectWithOptions(&options);
```

and then the second variant with one more step `acc = init(options); connect(acc)`.

Author: Oh, this part doesn't seem finished, there are a lot of empty lines too :(  I will update.
